enarx-keep-sev-shim/src/frame_allocator.rs

```diff
@@ -0,0 +1,192 @@
+        self.next_2M += Offset::from_items(1);
+
+        if self.next_2M.raw() > self.max {
+            return None;
+        }
+
+        Some(PhysFrame::containing_address(PhysAddr::new(
+            frame_address.raw() as _,
+        )))
+    }
+}
+
+/// Allocate memory and map it to the given virtual address
+pub fn allocate_and_map_memory(
+    map_to: VirtAddr,
+    size: usize,
+    mapper: &mut OffsetPageTable,
+    frame_allocator: &mut BootInfoFrameAllocator,
+) -> Result<&'static mut [u8], ()> {
+    #[inline(always)]
+    pub fn do_allocate_and_map_memory<S: PageSize>(
+        map_to: &[Page<S>],
+        mapper: &mut impl Mapper<S>,
+        frame_allocator: &mut (impl FrameAllocator<S> + FrameAllocator<Size4KiB>),
+    ) -> Result<(), ()> {
+        if map_to.len() == 0 {
+            return Ok(());
+        }
+        let size = map_to.len() * (Page::<S>::SIZE as usize);
+
+        let page_range = {
+            let heap_start = VirtAddr::from_ptr(map_to.as_ptr());
+            let heap_end = heap_start + size - 1u64;
+            let heap_start_page = Page::<S>::containing_address(heap_start);
+            let heap_end_page = Page::<S>::containing_address(heap_end);
+            Page::range_inclusive(heap_start_page, heap_end_page)
+        };
+
+        for page in page_range {
+            let frame = frame_allocator.allocate_frame().ok_or(())?;
+
+            let flags = PageTableFlags::PRESENT | PageTableFlags::WRITABLE;
+            unsafe {
+                mapper
+                    .map_to(page, frame, flags, frame_allocator)
+                    .map_err(|_| ())?
+                    .flush();
+            }
+        }
+        Ok(())
+    }
+
+    if size == 0 {
+        return Err(());
+    }
+
+    if !map_to.is_aligned(Page::<Size4KiB>::SIZE) {
+        return Err(());
+    }
+
+    let size = align_up(size as u64, Page::<Size4KiB>::SIZE) as usize;
+
+    let slice = unsafe {
+        core::slice::from_raw_parts_mut(map_to.as_mut_ptr::<Page4KiB>(), size / Page4KiB::size())
+    };
+
+    let (pre, middle, post) = unsafe { slice.align_to_mut::<Page2MiB>() };
+
+    let pre: &mut [Page<Size4KiB>] = unsafe {
+        core::slice::from_raw_parts_mut(
+            core::mem::transmute::<_, *mut Page<Size4KiB>>(pre.as_mut_ptr()),
+            pre.len(),
+        )
+    };
+
+    let middle: &[Page<Size2MiB>] = unsafe {
+        core::slice::from_raw_parts(
+            core::mem::transmute::<_, *const Page<Size2MiB>>(middle.as_ptr()),
+            middle.len(),
+        )
+    };
+
+    let post: &[Page<Size4KiB>] = unsafe {
+        core::slice::from_raw_parts(
+            core::mem::transmute::<_, *const Page<Size4KiB>>(post.as_ptr()),
+            post.len(),
+        )
+    };
+
+    do_allocate_and_map_memory(&pre, mapper, frame_allocator)?;
+    do_allocate_and_map_memory(&middle, mapper, frame_allocator)?;
+    do_allocate_and_map_memory(&post, mapper, frame_allocator)?;
```

Comment:
I think it's worth a comment here explaining that this is trying to allocate and map using the fewest number of frames possible.

Reply:
done